### PR TITLE
Add support for pools, users, CRUSH rules and EC profiles

### LINKED ISSUE #35: Add support for managing pools

### PR DESCRIPTION
**Pull request by cityofships**
Closes #35 

TODO: CI testing of the modules

### CODE REVIEW

File: plugins/module_utils/cephadm_common.py [hunk 1]
```diff
@@ -0,0 +1,78 @@
+#!/usr/bin/python3
+
+# Copyright 2020, Red Hat, Inc.
+# Copyright 2021, StackHPC, Ltd.
```
Comment (markgoddard) [inline plugins/module_utils/cephadm_common.py:4]: While it's good to keep the RedHat copyright, we should probably mention where these files were adapted from.

File: roles/ec_profiles/defaults/main.yml [hunk 2]
```diff
@@ -0,0 +1 @@
+cephadm_crush_rules: []
```
Comment (markgoddard) [inline roles/ec_profiles/defaults/main.yml:1]: ec_profiles

File: roles/keys/README.md [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+
+Optional groups (those services will be deployed when group exists)::
+
+* `rgws`
+
+All Ceph hosts must be in the `ceph` group.
+
+## Role variables
+
+* `cephadm_keys`: A list of pools to define
```
Comment (markgoddard) [inline roles/keys/README.md:29]: s/pools/keys/

**Review by markgoddard — CHANGES_REQUESTED** [review]
I stopped reviewing, lots of files need updating

File: roles/pools/README.md [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+* Either direct Internet access or private registry with desired Ceph image accessible to all hosts is required.
+
+### Inventory
+
+This role assumes the existence of the following groups:
+
+* `ceph`
+* `mons`
+* `mgrs`
+* `osds`
```
Comment (markgoddard) [inline roles/pools/README.md:19]: Just `mons`

File: roles/pools/README.md [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+
+Optional groups (those services will be deployed when group exists)::
+
+* `rgws`
+
+All Ceph hosts must be in the `ceph` group.
```
Comment (markgoddard) [inline roles/pools/README.md:25]: Delete

File: roles/pools/README.md [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+
+## Role variables
+
+* `cephadm_pools`: A list of pools to define
+   Example:
+   ```
+          cephadm_pools:
+            - name: pool1
+            - name: pool2
+              state: absent
```
Comment (markgoddard) [inline roles/pools/README.md:35]: Are these valid?

File: roles/keys/README.md [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+* The role assumes target hosts connection over SSH with user that has passwordless sudo configured.
+* Either direct Internet access or private registry with desired Ceph image accessible to all hosts is required.
+
+### Inventory
+
+This role assumes the existence of the following groups:
+
+* `mons`
+
+All Ceph hosts must be in the `ceph` group.
```
Comment (markgoddard) [inline roles/keys/README.md:18]: Not really

File: roles/ec_profiles/README.md [hunk 8]
```diff
@@ -0,0 +1,39 @@
+# keys
```
Comment (markgoddard) [inline roles/ec_profiles/README.md:1]: Whole readme is wrong

File: plugins/modules/cephadm_crush_rule.py [hunk 9]
```diff
@@ -0,0 +1,232 @@
...
+
+import datetime
+import json
+
+
+DOCUMENTATION = r'''
+---
+module: cephadm_crush_rule
+short_description: Manage Ceph Crush Replicated/Erasure Rule
+version_added: "1.6.0"
```
Comment (markgoddard) [inline plugins/modules/cephadm_crush_rule.py:34]: 1.4.0?

File: plugins/modules/cephadm_ec_profile.py [hunk 10]
```diff
@@ -0,0 +1,227 @@
...
+import json
+
+
+DOCUMENTATION = '''
+---
+module: cephadm_ec_profile
+
+short_description: Manage Ceph Erasure Code profile
+
+version_added: "1.6.0"
```
Comment (markgoddard) [inline plugins/modules/cephadm_ec_profile.py:36]: 1.4.0

File: plugins/modules/cephadm_ec_profile.py [hunk 11]
```diff
@@ -0,0 +1,227 @@
...
+        choices: ['present', 'absent', 'info']
+        default: present
+    stripe_unit:
+        description:
+            - The amount of data in a data chunk, per stripe.
+        required: false
+    k:
+        description:
+            - Number of data-chunks the object will be split in
+        required: true
```
Comment (markgoddard) [inline plugins/modules/cephadm_ec_profile.py:59]: false

File: plugins/modules/cephadm_ec_profile.py [hunk 12]
```diff
@@ -0,0 +1,227 @@
...
+    m:
+        description:
+            - Compute coding chunks for each object and store them on different
+              OSDs.
+        required: true
```
Comment (markgoddard) [inline plugins/modules/cephadm_ec_profile.py:64]: false

File: plugins/modules/cephadm_key.py [hunk 13]
```diff
@@ -0,0 +1,488 @@
...
+import time
+import base64
+
+DOCUMENTATION = r'''
+---
+module: cephadm_key
+
+author: Sebastien Han <seb@redhat.com>
+        Michal Nasiadka <michal@stackhpc.com>
+version_added: "1.6.0"
```
Comment (markgoddard) [inline plugins/modules/cephadm_key.py:38]: 1.4.0

File: plugins/modules/cephadm_key.py [hunk 14]
```diff
@@ -0,0 +1,488 @@
...
+            - The key output format when retrieving the information of an
+              entity.
+        required: false
+        default: json
+'''
+
+EXAMPLES = '''
+keys_to_create:
+  - name: client.key
+    secret: "AQAin8tUUK84ExAA/QgBtI7gEMWdmnvKBzlXdQ=="
```
Comment (markgoddard) [inline plugins/modules/cephadm_key.py:98]: Not used?

File: plugins/modules/cephadm_key.py [hunk 15]
```diff
@@ -0,0 +1,488 @@
...
+  with_items: "{{ keys_to_create }}"
+
+- name: delete cephx key
+  ceph_key:
+    name: "my_key"
+    state: absent
+
+- name: info cephx key
+  ceph_key:
+    name: "my_key""
```
Comment (markgoddard) [inline plugins/modules/cephadm_key.py:128]: Double double quote

**Review by markgoddard — CHANGES_REQUESTED** [review]

File: roles/pools/README.md (hunk 4 above)
Comment (mnasiadka) [reply roles/pools/README.md:19]: Done

File: plugins/modules/cephadm_key.py (hunk 14 above)
Comment (mnasiadka) [reply plugins/modules/cephadm_key.py:98]: Used in L111?

File: roles/pools/README.md (hunk 6 above)
Comment (mnasiadka) [reply roles/pools/README.md:35]: Updated.

**Review by mnasiadka — APPROVED** [review]

**Pull request merged**